### DISCUSSION TITLE
Castable attributes are omited on Realtime Validation

### DISCUSSION THREAD

**Discussion by DarkGhostHunter** [livewire/livewire#4814] [7. Bugs (archived)]
## What?

A model that uses Realtime validation with a casting attribute not present in the attributes array will error.

## Reproduction

1. Extend the User model.
2. Add a cast not present in the `users` table, like `['foo']`.
3. Create a Realtime Validation with an input of `foo`.

This is replicated in this repository: [https://github.com/noquierouser/laragear-rut-livewire-bug](https://github.com/noquierouser/laragear-rut-livewire-bug).

## Fix

Instead of getting the attributes of a model as an array, it should also check if the model has casts for  an specific attribute.

**Comment by joshhanley** [comment]
@DarkGhostHunter could you share your code as snippets here, which reproduces the issue with as little code as possible so we can investigate it? Thanks!

**Reply by DarkGhostHunter** [reply]
The model has a cast called `foo`, which doesn't exists as a column or property. Is an accessor/mutator for `foo_bar` and `foo_baz`.

```php
class Person extends Model
{
    $casts = [
        'foo' => CastsFooIntoFooBarAndFooBaz::class
    ]

    // ...
}
```

Like this cast:
```php
public function get($mode, $key, $value, $attributes)
{
    return $attributes['foo_bar'] . $attributes['foo_baz'];
}

public function get($mode, $key, $value, $attributes)
{
    $attributes['foo_bar'] = Str::afterLast($value, '_');
    $attributes['foo_baz'] = Str::beforeLast($value, '_');

    return $attributes;
}
```

In the Component using [realtime validation](https://laravel-livewire.com/docs/2.x/input-validation#real-time-validation), we validate the `foo` property directly:

```php
namespace App\Http\Livewire\Person;

use App\Models\Person;
use Livewire\Component;

class Create extends Component
{
    public Person $person;

    protected array $rules = [
        'person.foo' => ['required', 'string'],
        'person.name' => ['required'],
        'person.email' => ['required', 'email'],
    ];

    public function mount(Person $person)
    {
        $this->person = $person;
    }

    public function updated($propertyName)
    {
        $this->validateOnly($propertyName);
    }

    // ...
}
```

Doing the realtime validation in the frontend will return that the `foo` key doesn't exists. This is because the `foo` is a _virtual attribute_ that is never present in the model attributes. When is retrieved, the model is transformed to `toArray()` and then compared to the data to validate, hence the difference.

The solution should be to check if the cast attribute exists [before filtering the key to validate](https://github.com/livewire/livewire/blob/master/src/ComponentConcerns/ValidatesInput.php#L297). Otherwise that means appending a property that may be hidden from serialization, which may expose sensible data.

**Comment by joshhanley** [comment]
@DarkGhostHunter I personally I think if you want an attribute to exist, then appending it to the model is ok, as Livewire makes all properties of a model that have a rule, accessible to the front end anyway. But if you think otherwise, then PR's are welcome! 🙂
